### ISSUE TITLE
Get "Not a terminal" and signal SIGTTOU

### ISSUE THREAD

**Issue by HansH111** [ksh93/ksh#320]
We compiled yesterday the latest ksh93u+m on AIX
  so:   version         sh (AT&T Research) 93u+m/1.1.0-alpha 2021-08-13
  and tested if we could replace the current ksh93 (93t+ 2009-05-01)

We ran into problems when we tried to run interactively some scripts.
The scripts will hang  and after some investigation we found some parts which were causing the problem.

Basically what it does does is:
sudo -u  user1  /tmp/loadenv.sh which somexe
and the /tmp/loadenv.sh is checking which config files are present for the current user and loading them And we see that it starts to load a config with  . ./.settings.sh
and on that moment it hangs and you get the messages

Not a terminal
and 
./test.sh: line 3: 24510710: Stopped (SIGTTOU)

But the original scripts are also running from the crontab and they just work fine.

So it looks like this is something related to interactive and non interactive usage,....
Any hints before I try to reproduce this further ?

**Comment by HansH111** [comment]
I cannot reproduce this on Linux with same build or older build.

But on AIX I get the "Not a terminal" and the SIGTTOU  and the script hangs

2 scripts:
test.sh
```
#!/usr/local/bin/ksh93um
echo "starting"
notfound=$(sudo -u user2 /home/user1/testwrap which ls 2>&1 | tee 1.1 |grep -c "no ls ")
if [ $notfound -ne 0 ] ; then
   echo "ERR;ls not found"
else
   echo "OK-;ls found"
fi
exit 0
#EOF

```
testwrap
```
#!/usr/local/bin/ksh93um
usernm=$(whoami)
cd ~${usernm}
export HOME=$(pwd)
export TERM=dumb
echo "cmdline: $@"
exec $@
#EOF

```

2 users   user1 and user2,  
user1 starts test.sh
- test.sh will use sudo to switch to user2 and run the testwrap script
- as soon as it reached exec $@  it shows the error and signal


if I change the 1st line of the testwrap script  with #!/usr/bin/ksh93
it works correctly

if I remove the :  export HOME=$(pwd)   line with ksh93um
I don't get an error and it works correctly.

Hope this helps.

**Comment by McDutchie** [comment]
Unfortunately I cannot help by trying to reproduce and fix your problem on AIX. The hard disk on the AIX system at polarhome.com crashed and the maintainer does not have time to reinstall it with a new disk, so I am without access to any AIX system, indefinitely.

The string `Not a terminal` does not occur in the ksh93 source so it looks like some other command must have issued that message.

Have you tried ksh 93u+ 2012-08-01? Does the problem occur in that as well?

> if I remove the : export HOME=$(pwd) line with ksh93um
> I don't get an error and it works correctly.

Then try removing that and then in test.sh use `sudo -iu user2` instead of `sudo -u user2`. The `-i` option causes `sudo` to invoke a login shell which sets up variables like `HOME` correctly.

**Comment by HansH111** [comment]
yes, i know about polarhome.com....  I have the same problem. so i rented for 1 day a AIX 7.1 LPAR from siteox just to compile the latest ksh93u+m version

The problem is that for existing scripts it won't be a drop in replacement


I simplified the testwrap script a bit:
```
#!/usr/local/bin/ksh93t
#!/usr/local/bin/ksh93u
#!/usr/local/bin/ksh93um
export TERM=dumb
cd ~${USER}
export HOME=$PWD
echo "ksh version ${.sh.version}   cmdline: $@"
exec $@
#EOF
```
and each output:
```
user1> ./test.sh
OK-;ls found
user1>cat 1.1
ksh version Version M 93t+ 2009-05-01   cmdline: which ls
Not a terminal
/usr/bin/ls
```

```
user1> ./test.sh
OK-;ls found
user1>cat 1.1
ksh version Version AJM 93u+ 2012-08-01   cmdline: which ls
Not a terminal
/usr/bin/ls
```

```
user1> ./test.sh
./test.sh: line 5: 22020334: Stopped (SIGTTOU)
^Cuser1>cat 1.1
ksh version Version J 93u+m/1.1.0-alpha 2021-08-13   cmdline: which ls
Not a terminal
```

All ksh version give back "Not a terminal" but the 93u+m hangs because it receives a SIGTTOU signal.
So it looks like the problem lies not in the "Not a terminal" but somehow the signal causes the executing to halt

**Comment by HansH111** [comment]
just build a ksk93u+m/1.0.0-alpha 2021-01-24

```
user1> ./test.sh
OK-;ls found
user1> cat 1.1
ksh version Version AJM 93u+m/1.0.0-alpha 2021-01-24  cmdline: which ls
Not a terminal
/usr/bin/ls

```
that one works !
so between version 2021-01-24  and 2021-02-14  something is causing this problem.

**Comment by HansH111** [comment]
last version that worked correctly was on 10-feb-2021
https://github.com/ksh93/ksh/commit/2996d7ae7c86d763e48b269e972fe85567acede7

Next commit on 11-feb-2021 caused the SIGTTOU problems for AIX

**Comment by McDutchie** [comment]
Thank you for tracing that.

The next commit, which you say broke it, is: 37a18bab7173427ce045c9a67772cb98eb678cb7

That commit has one small change that fixed a really serious bug in which using shared-state command substitutions causes the shell to lose track of all running jobs. By any chance, does your script use any such command substitutions of the form `${ … }`?

To me, this looks like a case of a script that relied on a bug. ksh 93u+m is never going to be a drop-in replacement for scripts that rely on bugs, I'm afraid.

Having said that...

SIGTTOU is issued by the system when a background job tries to write to the terminal. By default, it stops the process. That seems like it is probably the cause of the hang.

However, you should be able to disable SIGTTOU by issuing `stty -tostop`, thus allowing background processes to write to the terminal. (Re-enable it with `stty tostop`).

Does that fix the hang?

**Comment by McDutchie** [comment]
Closing due to inactivity. @HansH111, please do let me know if the idea in the previous comment fixes your problem.

**Comment by HansH111** [comment]
> By any chance, does your script use any such command substitutions of the form ${ … }?
no it doesn't

> However, you should be able to disable SIGTTOU by issuing stty -tostop, thus allowing background processes to write to the terminal. (Re-enable it with stty tostop).
No the stty -tostop does not solve the problem, it makes no
difference, maybe this has something to do with sudo

The only thing which solves the problem seems to be this small change:
in the testwrap script
place an "unset HOME"  before the "export HOME=$PWD"
And then it works....
or remove the exporting of HOME
Seems an assignment of HOME does something special and triggers the error...

**Comment by JohnoKing** [comment]
> Seems an assignment of HOME does something special and triggers the error...

That is a bit strange, although it does give me an idea. Perhaps ksh is getting the value of `$HOME` from the wrong scope. I have made a patch that makes ksh consistently use `sh_scoped` when obtaining `$HOME`. Does this fix the error?
```diff
--- a/src/cmd/ksh93/bltins/cd_pwd.c
+++ b/src/cmd/ksh93/bltins/cd_pwd.c
@@ -112,7 +112,7 @@ int	b_cd(int argc, char *argv[],Shbltin_t *context)
 	if(argc==2)
 		dir = sh_substitute(oldpwd,dir,argv[1]);
 	else if(!dir)
-		dir = nv_getval(HOME);
+		dir = nv_getval(sh_scoped(shp,HOME));
 	else if(*dir == '-' && dir[1]==0)
 		dir = nv_getval(opwdnod);
 	if(!dir || *dir==0)
--- a/src/cmd/ksh93/edit/history.c
+++ b/src/cmd/ksh93/edit/history.c
@@ -215,7 +215,7 @@ int  sh_histinit(void *sh_context)
 	if(!(histname = nv_getval(HISTFILE)))
 	{
 		int offset = staktell();
-		if(cp=nv_getval(HOME))
+		if(cp=nv_getval(sh_scoped(shp,HOME)))
 			stakputs(cp);
 		stakputs(hist_fname);
 		stakputc(0);
--- a/src/cmd/ksh93/sh/init.c
+++ b/src/cmd/ksh93/sh/init.c
@@ -1311,7 +1311,7 @@ Shell_t *sh_init(register int argc,register char *argv[], Shinit_f userinit)
 	save_envmarker = env_init(shp);
 	if(!ENVNOD->nvalue.cp)
 	{
-		sfprintf(shp->strbuf,"%s/.kshrc",nv_getval(HOME));
+		sfprintf(shp->strbuf,"%s/.kshrc",nv_getval(sh_scoped(shp,HOME)));
 		nv_putval(ENVNOD,sfstruse(shp->strbuf),NV_RDONLY);
 	}
 	*SHLVL->nvalue.ip +=1;
```

**Comment by HansH111** [comment]
No, that didn't help, but I found out something else...

#!/usr/local/bin/kshbeta2
cd ~$USER
#unset HOME
HOME=$PWD
echo "ksh version ${.sh.version}  home=$HOME $PWD cmdline: $@"
exec  $@

The user2 has a default /usr/bin/csh shell,  so when you can see the output
and then after that it hangs or give that SIGTTOU error
But when you look in the process list you see, that with the exec it starts
a /usr/bin/csh which starts an stty  and then it hangs

 user2 40894664 38011004   0 14:52:01  pts/1  0:00 /usr/bin/csh -f
/usr/bin/which ls
 user2 37814332 40894664   0 14:52:02  pts/1  0:00 stty erase ^h kill ^u
intr ^c susp ^z

when I remove the # before the unset,  it works
if I do an export HOME=$PWD   it always fails

**Comment by McDutchie** [comment]
Since you seem to have found a workaround and I do not believe the original problem was due to a bug in ksh, I think there's nothing further for us to do here.

**Comment by HansH111** [comment]
Hi Martijn,

Still having this problem on AIX only
So I spent the last few days in building different versions to find out
which commit is causing the problem on AIX.
because I really want to use the latest ksh93u+m version...

commit 37a18bab7173427ce045c9a67772cb98eb678cb7
Last working version which returned complete output, so:
ksh version Version AJM 93u+m/1.0.0-alpha 2021-02-11  doing: exec which ls
/usr/bin/ls

commit 41ebb55a3a9133f58cf3cc1c47e83bb5dc11404e
Instead of getting the output, I get as output
ksh version Version AJM 93u+m/1.0.0-alpha 2021-02-11  doing: exec which ls

Background write


Seems that the change in src/cmd/ksh93/sh/jobs.c is causing this...
if (!sh_isoption(SH_INTERACTIVE))
    return;
The above was removed from job_init function , but if I put it back again,
it works again...
Even if I do that change in version 1.0.2  it works again
I don't get a SIGTTOU or 'Background write' anymore..

Probably I broke that way the complete job control functionality...
Any suggestions ?

Hans

**Comment by McDutchie** [comment]
Hi @HansH111,

I now have access to an AIX testing account. But without a reproducer script I have no real chance to figure out what's going on. Could you post a reproducer that I can actually run, please?

**Comment by McDutchie** [comment]
> Seems that the change in src/cmd/ksh93/sh/jobs.c is causing this... if (!sh_isoption(SH_INTERACTIVE)) return; The above was removed from job_init function , but if I put it back again, it works again... Even if I do that change in version 1.0.2 it works again I don't get a SIGTTOU or 'Background write' anymore.. Probably I broke that way the complete job control functionality...

I have been testing ksh with those lines restored. In fact it doesn't seem like it breaks anything. I made some rather extensive regression tests for job control in scripts and they all pass. Everything about job_init() after those restored lines is evidently to do with job control on interactive shells, which, upon rereading the code, also makes sense to me now.

Would still be nice to have a reproducer for your problem but it may not be as essential now.

**Comment by McDutchie** [comment]
@HansH111: Hopefully your problem is now solved in the 1.0.3 release. Please let me know.

**Comment by HansH111** [comment]
Indeed it is..... with the 1.0.3 release, the problem is solved

thx
Hans
